### ISSUE TITLE
Skip filtering of CTE in some cases

### ISSUE THREAD

**Issue by 80LevelElf** [linq2db/linq2db#1564]
Hi!
Let's just imagine that I have to get the list of all categories in my product store and each category should display the full path to itself like: Gadgets -> Phones -> Apple

The simplest solution is to use CTE to get the tree of all parents of some category like:

```
        private IQueryable<AdminCategoryPathItemCte> GetPathQuery(IStoreDb db)
        {
            var categoryPathCte = db.GetCte<AdminCategoryPathItemCte>(categoryHierarchy =>
            {
                return
                    (
                        from innerC in db.Categories
                        select new AdminCategoryPathItemCte
                        {
                            CategoryId = innerC.Id,
                            ParentCategoryId = innerC.ParentId,
                            Name = innerC.Name,
                            RootCategoryId = innerC.Id,
                            Level = 0
                        }
                    )
                    .Concat
                    (
                        from c in db.Categories
                        from ch in categoryHierarchy.InnerJoin(ch => ch.ParentCategoryId == c.Id)
                        select new AdminCategoryPathItemCte
                        {
                            CategoryId = c.Id,
                            ParentCategoryId = c.ParentId,
                            Name = c.Name,
                            RootCategoryId = ch.RootCategoryId,
                            Level = ch.Level + 1
                        }
                    );
            });

            return categoryPathCte;
        }
```

So the Apple category will have 0 Level, Phones  - 1 level and Gadgets  - 2 level.
But we need the reverse order, right?

Let's make it this way:
`var pathQuery = GetPathQuery(db).OrderByDescending(i => i.Level);`

If we will see the query of the 'pathQuery' variable we will see:
```
--  PostgreSQL.9.5 PostgreSQL
WITH RECURSIVE categoryHierarchy
(
	RootCategoryId, 
	CategoryId, 
	ParentCategoryId, 
	Name, 
	"Level"
)
AS
(
	SELECT
		innerC.Id,
		innerC.Id,
		innerC.ParentId,
		innerC.Name,
		0
	FROM
		Category innerC
	UNION ALL
	SELECT
		ch.RootCategoryId,
		c_1.Id,
		c_1.ParentId,
		c_1.Name,
		ch."Level" + 1
	FROM
		Category c_1
			INNER JOIN categoryHierarchy ch ON ch.ParentCategoryId = c_1.Id
)
SELECT
	t1.RootCategoryId,
	t1.ParentCategoryId,
	t1.Name,
	t1."Level"
FROM
	categoryHierarchy t1
**ORDER BY
	t1."Level" DESC**
```

There is orderBy term at this moment.

But after we will try to use such CTE in other query:
```
var adminCategoriesQuery =
     from c in db.Categories
     select new AdminCategoryPreview
     {
         Id = c.Id,
         IsVisible = c.IsVisible,
         DisplayOrder = c.DisplayOrder,
         FullPath = pathQuery.Where(c1 => c1.RootCategoryId == c.Id)
					.StringAggregate(" -> ", i => i.Name)
     };
```
the ordering term will be skipped:
```
--  PostgreSQL.9.5 PostgreSQL
WITH RECURSIVE categoryHierarchy
(
	RootCategoryId, 
	CategoryId, 
	ParentCategoryId, 
	Name, 
	"Level"
)
AS
(
	SELECT
		innerC.Id,
		innerC.Id,
		innerC.ParentId,
		innerC.Name,
		0
	FROM
		Category innerC
	UNION ALL
	SELECT
		ch.RootCategoryId,
		c_1.Id,
		c_1.ParentId,
		c_1.Name,
		ch."Level" + 1
	FROM
		Category c_1
			INNER JOIN categoryHierarchy ch ON ch.ParentCategoryId = c_1.Id
)
SELECT
	c_2.Id,
	c_2.IsVisible,
	c_2.DisplayOrder,
	(
		SELECT
			STRING_AGG(c1.Name, ' -> ')
		FROM
			categoryHierarchy c1
		WHERE
			c1.RootCategoryId = c_2.Id
	)
FROM
	Category c_2
```

Looks like the second query use the root CTE, not the needed select with ordering.
So the ordering term is just skipped.

**Environment details**
linq2db version: 2.6.3
Database Server: PostgreSQL 10
Database Provider: standard provider "PostgreSQL.9.5"
Operating system: Windows 10
Framework version: .Net core 2.1

**Comment by sdanyliv** [comment]
@80LevelElf, just for saving my time. Does Posgres supports order by in subqueries?

**Comment by 80LevelElf** [comment]
Yes, Postgres supports order by in subqueries.
And I have a proposal:

```
	(
		SELECT
			STRING_AGG(c1.Name, ' -> ')
		FROM
			categoryHierarchy c1
		WHERE
			c1.RootCategoryId = c_2.Id
	)
```

I want to use my order by statement here. But it is an aggregation function.
So Postgres force me to use group by statement with order by statement and so on.
Maybe it is a reason Linq2Db skip ordering?

**Comment by sdanyliv** [comment]
Ok, write which SQL do you expect. I know place which removes ordering and can fix that.

**Comment by sdanyliv** [comment]
Also maybe you need this syntax?
```sql
	(
		SELECT
			STRING_AGG(c1.Name, ' -> ') WITHIN GROUP (ORDER BY c1."Level" DESC)
		FROM
			categoryHierarchy c1
		WHERE
			c1.RootCategoryId = c_2.Id
	)
```

**Comment by 80LevelElf** [comment]
Sorry for the delay! I was on my job.

I think the most sophisticated syntax in Postgres will be:

```
	(
		SELECT
			STRING_AGG(c1.Name, ' -> ' ORDER BY c1."Level" DESC)
		FROM
			categoryHierarchy c1
		WHERE
			c1.RootCategoryId = c_2.Id
	)
```
And looks like your sintax are not allowed in PG:
https://stackoverflow.com/questions/40666324/how-to-translate-within-group-from-oracle-to-in-postgres

I think that it is the best solution because it will work for all DBs:

```
	(
		SELECT
			STRING_AGG(c1.Name, ' -> ')
		FROM
			(select cInner.Name, cInner.RootCategoryId from categoryHierarchy cInner order by "Level" DESC) c1
		WHERE
			c1.RootCategoryId = c_2.Id
	)
```

What do you think?

**Comment by sdanyliv** [comment]
Ok, i think it is doable will prepare sample.

**Comment by 80LevelElf** [comment]
@sdanyliv hi!
Any updates?

**Comment by sdanyliv** [comment]
Sorry not yet. Leak of time. But you can easily create such implementation by yourself.
As an example you can look at https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Sql/Sql.Analytic.cs#L206
Here we define Window Functions.
So you can have such syntax `Sql.Ext.Aggregate(" -> ", t.SomeField).OrderBy(t.SomeValue).ToVaue()`
`TokenName` and `ChainPrecedence` is a key of exploration.

**Comment by MaceWindu** [comment]
Fixed by #1823
